GH-8626: Provide cleaner `transform()` DSL

artembilan opened the pull request:
Fixes https://github.com/spring-projects/spring-integration/issues/8626

* Add missed `transform(String beanName, @Nullable String methodName)` API
* Introduce a `TransformerSpec` to expose a strict API to configure transformer variants.
* Introduce `transformWith(Consumer<TransformerSpec>)` as a single point of all possible transformer and its endpoint options
* Deprecate those `IntegrationFlowDefinition.transform()` variants which are harder to configure as several lambda arguments

This change will make Kotlin & Groovy DSL more readable and straightforward

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-8626: Provide cleaner `transform()` DSL

Fixes https://github.com/spring-projects/spring-integration/issues/8626

* Add missed `transform(String beanName, @Nullable String methodName)` API
* Introduce a `TransformerSpec` to expose a strict API to configure
transformer variants.
* Introduce `transformWith(Consumer<TransformerSpec>)` as a single point
of all possible transformer and its endpoint options
* Deprecate those `IntegrationFlowDefinition.transform()` variants
which are harder to configure as several lambda arguments

This change will make Kotlin & Groovy DSL more readable and straightforward

artembilan commented:
@garyrussell ,

I have a naming problem 😄 

my new DSL method is `transformWith(Consumer<TransformerSpec> transformerConfigurer)`.
If I leave it just as `transform()`, the "typeless" lambda style configuration would clash with an existing `transform(GenericTransformer<S, T> genericTransformer)`.
Therefore only way to satisfy compiler is to give the method a different name.

So, is that `transformWith` OK? Or do you see any other possible names?

NOTE: The change in not done yet.
1. Revise tests in other modules.
2. Fix docs
3. The Kotlin & Groovy DSLs might be done in the separate PR.

I don't deprecate many other `transform()` methods because I call them shortcuts. It is really more straightforward just to type `.<String, String>transform(String::toUpperCase)`, then try something like `.transformWith(t -> t.<String, String>transformer(String::toUpperCase))`.
Or similar for just expression, or bean method call.
Probably for Kotlin & Groovy we won't have them since it looks like their DSL is more about options and their values rather then method calls.

Thanks

artembilan committed:
* Use new `transformWith()` in tests where a deprecated API still used
* Add JavaDocs to `TransformerSpec`
* Fix generic types for `BaseIntegrationFlowDefinition.transformWith()` -
they make sense exactly on the `TransformerSpec.transformer()` only
* Apply `transformWith()` for Groovy DSL
* Introduce a new `ClassUtils.isLambda(Object candidate)`
and add a check for Groovy `Closure`
* Fix `GroovyIntegrationFlowDefinition.createConfigurerIfAny()` to propagate a `Consumer` argument
down to the `Closure`

garyrussell commented:
`transformWith()` is fine for me. Should I wait for other changes before reviewing?

artembilan commented:
Yes, please.
I've already applied the change for Groovy DSL to justify it.
So, only Kotlin is left and docs.

Thanks

artembilan committed:
* Rename `TransformerSpec -> TransformerEndpointSpec` for better context
meaning of the class
* Introduce `KotlinTransformerEndpointSpec` as an extension of the `TransformerEndpointSpec`
to have an `inline fun <reified P> transformer(crossinline function: (P) -> Any)`
for Kotlin style
* Add `KotlinIntegrationFlowDefinition.transformWith(KotlinTransformerEndpointSpec)`
* Deprecate Kotlin methods which are covered by the mentioned `transformWith()`
* Fix tests to use new API
* Mentioned the change in the doc

garyrussell reviewed (approved):
LGTM; needs to be removed from Draft mode for merge.

garyrussell commented:
Removed from draft mode - let me know if that is not what you intended.

artembilan commented:
> Removed from draft mode

Yes. That is exactly what we need now.
Feel free to merge.
Thanks.
